### PR TITLE
fix(v2): refactor color mode system

### PR DESCRIPTION
**Pull request by slorber**
## Motivation

Some things impossible with current api:

- disable the toggle, but use the dark mode by default
- disable the toggle, and respect the user preferred color
- force default light theme, ignoring the user preferred color
- set dark theme by default, but respect the user preference to light if set

Breaking change to classic theme config to support all possible scenarios

Superseed #2939 (as I now work on a branch of the main repo)

## Doc

image

### CODE REVIEW

**Comment by isaacito** [comment]
Would love to have this!

File: packages/docusaurus-theme-classic/src/theme/hooks/useTheme.ts [hunk 1]
```diff
@@ -10,64 +10,72 @@ import {useState, useCallback, useEffect} from 'react';
...
   dark: 'dark',
 };
 
+// Ensure to always return a valid theme even if input is invalid
+const coerceToTheme = (theme) => {
+  return theme === themes.dark ? themes.dark : themes.light;
+};
+
+const getInitialTheme = () => {
+  if (typeof document === 'undefined') {
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/theme/hooks/useTheme.ts:23]: Use [`ExecutionEnvironment.canUseDOM()`](https://v2.docusaurus.io/docs/docusaurus-core/#executionenvironment)

File: website/docs/theme-classic.md [hunk 2]
```diff
@@ -11,32 +11,43 @@ This section is a work in progress.
...
+    },
     // ...
   },
+  // ...
 };
 ```
 
+:::caution
+
+With `respectPrefersColorScheme: true`, the `defaultMode` is overriden by user system preferences.
```
Comment (yangshun) [inline website/docs/theme-classic.md:45]: Typo - overridden.

File: packages/docusaurus-theme-classic/src/index.js [hunk 3]
```diff
@@ -20,41 +20,46 @@ const ContextReplacementPlugin = requireFromDocusaurusCore(
...
 
-  var darkQuery = window.matchMedia('(prefers-color-scheme: dark)');
-
-  var preferredTheme = getPreferredTheme();
-  if (preferredTheme !== null) {
-    setDataThemeAttribute(preferredTheme);
-  } else if (darkQuery.matches || defaultDarkMode) {
-    setDataThemeAttribute('dark');
+  var storedTheme = getStoredTheme();
+  if (storedTheme !== null) {
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/index.js:41]: I think we can format this part better. Perhaps paste this part into https://prettier.io/playground/ and copy the formatted contents back?

**Pull request merged**